File: esw-utils/src/main/scala/esw/highlevel/dsl/EventServiceDsl.scala
package esw.highlevel.dsl

import akka.Done
import akka.actor.Cancellable
import csw.event.api.scaladsl.{EventPublisher, EventService, EventSubscriber, EventSubscription}
import csw.params.core.generics.Parameter
import csw.params.core.models.Prefix
import csw.params.events._

import scala.concurrent.duration.FiniteDuration
import scala.concurrent.{ExecutionContext, Future}

trait EventServiceDsl {
  private[esw] def eventService: EventService
  private lazy val publisher: EventPublisher   = eventService.defaultPublisher
  private lazy val subscriber: EventSubscriber = eventService.defaultSubscriber

  def systemEvent(sourcePrefix: String, eventName: String, parameters: Parameter[_]*): SystemEvent =
    SystemEvent(Prefix(sourcePrefix), EventName(eventName), parameters.toSet)

  def observeEvent(sourcePrefix: String, eventName: String, parameters: Parameter[_]*): ObserveEvent =
    ObserveEvent(Prefix(sourcePrefix), EventName(eventName), parameters.toSet)

  def publishEvent(event: Event): Future[Done] = publisher.publish(event)

  def publishEvent(every: FiniteDuration)(eventGenerator: => Option[Event])(implicit ec: ExecutionContext): Cancellable =
    publisher.publishAsync(Future(eventGenerator), every)

  private val stringToEventKey = (x: String) => EventKey(x)
  def onEvent(eventKeys: String*)(callback: Event => Unit)(implicit ec: ExecutionContext): EventSubscription =
    subscriber.subscribeAsync(eventKeys.toSet.map(stringToEventKey(_)), event => Future(callback(event)))

  def getEvent(eventKeys: String*): Future[Set[Event]] = {
    val value: Set[EventKey] = eventKeys.toSet.map(stringToEventKey(_))
    subscriber.get(value)
  }
}

File: src/main/kotlin/esw/highlevel/dsl/EventServiceDsl.kt
package esw.highlevel.dsl

import akka.Done
import akka.actor.Cancellable
import csw.event.api.scaladsl.EventSubscription
import csw.params.core.generics.Parameter
import csw.params.events.Event
import csw.params.events.ObserveEvent
import csw.params.events.SystemEvent
import esw.highlevel.dsl.javadsl.JEventServiceDsl
import esw.ocs.macros.StrandEc
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.future.await
import kotlinx.coroutines.future.future
import scala.concurrent.duration.FiniteDuration
import java.time.Duration
import java.util.*
import java.util.concurrent.CompletionStage
import java.util.concurrent.TimeUnit
import java.util.function.Supplier

interface EventServiceDsl : CoroutineScope, JEventServiceDsl {

    fun strandEc(): StrandEc

    fun systemEvent(sourcePrefix: String, eventName: String, vararg parameters: Parameter<*>): SystemEvent =
        jSystemEvent(sourcePrefix, eventName, parameters.toSet())

    fun observeEvent(sourcePrefix: String, eventName: String, vararg parameters: Parameter<*>): ObserveEvent =
        jObserveEvent(sourcePrefix, eventName, parameters.toSet())

    suspend fun publishEvent(event: Event): Done = jPublishEvent(event).await()

    // todo: see if this works and simplify
    suspend fun publishEvent(every: Duration, eventGenerator: suspend () -> Optional<Event>): Cancellable {
        val xx: Supplier<CompletionStage<Optional<Event>>> = coroutineScope {
            Supplier {
                future {
                    eventGenerator()
                }.minimalCompletionStage()
            }
        }

        return jPublishEventAsync(every, xx, strandEc())
    }

    fun onEvent(vararg eventKeys: String, callback: (Event) -> Unit): EventSubscription =
        jOnEvent(eventKeys.toSet(), callback, strandEc())

    suspend fun getEvent(vararg eventKeys: String): Set<Event> =
        jGetEvent(eventKeys.toSet(), strandEc()).await().toSet()

}
